File: src/main/java/io/scriptor/http/HTTPResult.java
package io.scriptor.http;

import java.io.InputStream;

public abstract class HTTPResult<T> {

    private final int statusCode;
    private final String statusText;

    private final T value;

    public HTTPResult(final int statusCode) {
        this.statusCode = statusCode;
        this.statusText = null;
        this.value = null;
    }

    public HTTPResult(final int statusCode, final String statusText) {
        this.statusCode = statusCode;
        this.statusText = statusText;
        this.value = null;
    }

    public HTTPResult(final int statusCode, final String statusText, final T value) {
        this.statusCode = statusCode;
        this.statusText = statusText;
        this.value = value;
    }

    @Override
    public String toString() {
        return "Result( statusCode=%d, statusText=%s, value=%s )".formatted(statusCode, statusText, value);
    }

    public int getStatusCode() {
        return statusCode;
    }

    public String getStatusText() {
        return statusText;
    }

    public T getValue() {
        return value;
    }

    public abstract int getSize();

    public abstract InputStream getStream();
}

File: server/src/main/kotlin/dev/scriptor/server/http/result/HTTPResult.kt
package dev.scriptor.server.http.result

import dev.scriptor.server.http.ParameterList
import java.nio.channels.ReadableByteChannel

abstract class HTTPResult<T> {

    val statusCode: Int
    val statusText: String

    val headers: ParameterList

    val position: Long
    val count: Long

    val channel: ReadableByteChannel?

    constructor(
        statusCode: Int = 200,
        statusText: String = "OK",
        headers: ParameterList = ParameterList(),
        position: Long = 0L,
        count: Long = -1L,
        channel: ReadableByteChannel? = null,
    ) {
        this.statusCode = statusCode
        this.statusText = statusText
        this.headers = headers
        this.position = position
        this.count = count
        this.channel = channel
    }
}
